### PR TITLE
feat: impl pop_last() & pop_first() for BTreeMap

### PR DESCRIPTION
**Pull request by witter-deland**
BtreeMap will support deque after this PR #106.
And more methods for BTreeMap for #87

### CODE REVIEW

File: src/btreemap.rs [hunk 1]
```diff
@@ -1374,6 +1406,32 @@ mod test {
...
+            assert_eq!(btree.insert(b(&[]), b(&[])), None);
+            assert_eq!(btree.allocator.num_allocated_chunks(), 1);
+
+            assert_eq!(btree.pop_last(), Some((b(&[]), b(&[]))));
+            assert_eq!(btree.allocator.num_allocated_chunks(), 0);
+        });
+    }
+
+    #[test]
+    fn allocations_4() {
```
Comment (ielashi) [inline src/btreemap.rs:1423]:
```suggestion
    fn pop_first_single_entry() {
```

File: src/btreemap.rs [hunk 2]
```diff
@@ -1374,6 +1406,32 @@ mod test {
...
+    #[test]
+    fn allocations_3() {
+        btree_test(|mut btree| {
+            assert_eq!(btree.allocator.num_allocated_chunks(), 0);
+
+            assert_eq!(btree.insert(b(&[]), b(&[])), None);
+            assert_eq!(btree.allocator.num_allocated_chunks(), 1);
+
+            assert_eq!(btree.pop_last(), Some((b(&[]), b(&[]))));
+            assert_eq!(btree.allocator.num_allocated_chunks(), 0);
```
Comment (ielashi) [inline src/btreemap.rs:1418]: nit: I'd also assert that `is_empty` returns true.

File: src/btreemap.rs [hunk 3]
```diff
@@ -1374,6 +1406,32 @@ mod test {
...
+    #[test]
+    fn allocations_4() {
+        btree_test(|mut btree| {
+            assert_eq!(btree.allocator.num_allocated_chunks(), 0);
+
+            assert_eq!(btree.insert(b(&[]), b(&[])), None);
+            assert_eq!(btree.allocator.num_allocated_chunks(), 1);
+
+            assert_eq!(btree.pop_first(), Some((b(&[]), b(&[]))));
+            assert_eq!(btree.allocator.num_allocated_chunks(), 0);
```
Comment (ielashi) [inline src/btreemap.rs:1431]: nit: I'd also assert that `is_empty` returns true.

File: src/btreemap.rs [hunk 4]
```diff
@@ -1494,6 +1552,26 @@ mod test {
         });
     }
 
+    #[test]
+    fn pop_last_simple() {
```
Comment (ielashi) [inline src/btreemap.rs:1556]: Can we also add separate tests where we call `pop_last` and `pop_first` on an empty btree?

File: src/btreemap.rs [hunk 5]
```diff
@@ -1938,6 +2016,86 @@ mod test {
...
+
+        for j in 0..=10 {
+            for i in 0..=255 {
+                assert_eq!(btree.get(&b(&[i, j])), None);
+                assert_eq!(std_btree.get(&b(&[i, j])), None);
+            }
+        }
+
+        // We've deallocated everything.
+        assert_eq!(btree.allocator.num_allocated_chunks(), 0);
```
Comment (ielashi) [inline src/btreemap.rs:2056]: ditto (checking is empty)

File: src/btreemap.rs [hunk 6]
```diff
@@ -1938,6 +2016,86 @@ mod test {
         assert_eq!(btree.allocator.num_allocated_chunks(), 0);
     }
 
+    #[test]
+    fn many_insertions_3() {
```
Comment (ielashi) [inline src/btreemap.rs:2020]:
```suggestion
    fn pop_first_many_entries() {
```

File: src/btreemap.rs [hunk 7]
```diff
@@ -1938,6 +2016,86 @@ mod test {
...
+                assert_eq!(std_btree.get(&b(&[i, j])), None);
+            }
+        }
+
+        // We've deallocated everything.
+        assert_eq!(btree.allocator.num_allocated_chunks(), 0);
+    }
+
+    #[test]
+    fn many_insertions_4() {
```
Comment (ielashi) [inline src/btreemap.rs:2060]:
```suggestion
    fn pop_last_many_entries() {
```

File: src/btreemap.rs [hunk 8]
```diff
@@ -1938,6 +2016,86 @@ mod test {
...
+
+        for j in 0..=10 {
+            for i in 0..=255 {
+                assert_eq!(btree.get(&b(&[i, j])), None);
+                assert_eq!(std_btree.get(&b(&[i, j])), None);
+            }
+        }
+
+        // We've deallocated everything.
+        assert_eq!(btree.allocator.num_allocated_chunks(), 0);
```
Comment (ielashi) [inline src/btreemap.rs:2096]: ditto (check `is_empty`)

File: src/btreemap.rs [hunk 9]
```diff
@@ -1990,6 +2148,44 @@ mod test {
         });
     }
 
+    #[test]
+    fn len_2() {
```
Comment (ielashi) [inline src/btreemap.rs:2152]:
```suggestion
    fn pop_first_len() {
```

File: src/btreemap.rs [hunk 10]
```diff
@@ -1990,6 +2148,44 @@ mod test {
...
+                assert_eq!(btree.pop_first().unwrap().1, b(&i.to_le_bytes()));
+            }
+
+            assert_eq!(btree.len(), 0);
+            assert!(btree.is_empty());
+        });
+    }
+
+    #[test]
+    fn len_3() {
```
Comment (ielashi) [inline src/btreemap.rs:2171]:
```suggestion
    fn pop_last_len() {
```

File: src/btreemap/proptests.rs [hunk 11]
```diff
@@ -196,5 +200,33 @@ fn execute_operation<M: Memory>(
...
+        Operation::PopLast => {
+            assert_eq!(std_btree.len(), btree.len() as usize);
+            if std_btree.is_empty() {
+                return;
+            }
+
+            let idx = std_btree.len();
+
+            if let Some(entry) = btree.iter().skip(idx).take(1).next() {
+                eprintln!("Pop last, the key is {}", hex::encode(&entry.0));
```
Comment (ielashi) [inline src/btreemap/proptests.rs:212]: These printlns are written in a special format to be easy to replay a failed test case later.

```suggestion
                eprintln!("PopLast");
```

File: src/btreemap/proptests.rs [hunk 12]
```diff
@@ -196,5 +200,33 @@ fn execute_operation<M: Memory>(
...
         }
+        Operation::PopLast => {
+            assert_eq!(std_btree.len(), btree.len() as usize);
+            if std_btree.is_empty() {
+                return;
+            }
+
+            let idx = std_btree.len();
+
+            if let Some(entry) = btree.iter().skip(idx).take(1).next() {
```
Comment (ielashi) [inline src/btreemap/proptests.rs:211]: I don't understand why we need this `if` statement. Is it not sufficient to compare `std_btree.pop_last()` with `btree.pop_last()` and check that they're equal?

File: src/btreemap/proptests.rs [hunk 13]
```diff
@@ -196,5 +200,33 @@ fn execute_operation<M: Memory>(
...
+        Operation::PopFirst => {
+            assert_eq!(std_btree.len(), btree.len() as usize);
+            if std_btree.is_empty() {
+                return;
+            }
+
+            let idx = std_btree.len();
+
+            if let Some(entry) = btree.iter().skip(idx).take(1).next() {
+                eprintln!("Pop first, the key is {}", hex::encode(&entry.0));
```
Comment (ielashi) [inline src/btreemap/proptests.rs:226]:
```suggestion
                eprintln!("PopFirst");
```

File: src/btreemap.rs [hunk 14]
```diff
@@ -570,6 +570,38 @@ where
...
     }
 
+    /// Removes and returns the last element in the map. The key of this element is the maximum key that was in the map
+    pub fn pop_last(&mut self) -> Option<(K, V)> {
+        if self.root_addr == NULL {
+            return None;
+        }
+
+        let root_node = self.load_node(self.root_addr);
+        root_node.max_key(self.memory()).map(|max_key| {
```
Comment (ielashi) [inline src/btreemap.rs:580]: Why don't we use `root_node.get_max` here?

File: src/btreemap.rs [hunk 15]
```diff
@@ -570,6 +570,38 @@ where
...
+    }
+
+    /// Removes and returns the first element in the map. The key of this element is the minimum key that was in the map
+    pub fn pop_first(&mut self) -> Option<(K, V)> {
+        if self.root_addr == NULL {
+            return None;
+        }
+
+        let root_node = self.load_node(self.root_addr);
+        root_node.min_key(self.memory()).map(|min_key| {
```
Comment (ielashi) [inline src/btreemap.rs:596]: ditto (we have `get_min`)

File: src/btreemap/proptests.rs (hunk 12 above)
Comment (witter-deland) [reply src/btreemap/proptests.rs:211]: get_max and remove call value reading twice, which is unnecessary. Especially when there are N items of data, this overhead will be * N

File: src/btreemap.rs (hunk 14 above)
Comment (witter-deland) [reply src/btreemap.rs:580]: get_max and remove call value reading twice, which is unnecessary. Especially when there are N items of data, this overhead will be * N

Comment (ielashi) [reply src/btreemap.rs:615]: Reading values are cheap, so I would keep the code simple by reusing the same code and only optimize it if a benchmark indicates that it would lead to a significant performance boost.

File: src/btreemap/proptests.rs [hunk 16]
```diff
@@ -34,6 +36,8 @@ fn operation_strategy() -> impl Strategy<Value = Operation> {
         15 => (any::<usize>()).prop_map(Operation::Remove),
         5 => (any::<usize>(), any::<usize>())
             .prop_map(|(from, len)| Operation::Range { from, len }),
+        15 =>  Just(Operation::PopFirst),
+        15 =>  Just(Operation::PopLast),
```
Comment (ielashi) [inline src/btreemap/proptests.rs:40]: nit: I'd give these smaller weights. We want the btree in this test grow over time, and putting large weights on removing elements will keep the btree too small to have meaningful tests.

```suggestion
        2 =>  Just(Operation::PopFirst),
        2 =>  Just(Operation::PopLast),
```

File: src/btreemap/proptests.rs [hunk 17]
```diff
@@ -231,5 +235,27 @@ fn execute_operation<M: Memory>(
...
+        Operation::PopLast => {
+            assert_eq!(std_btree.len(), btree.len() as usize);
+
+            let idx = std_btree.len();
+
+            if let Some(entry) = btree.iter().skip(idx).take(1).next() {
+                eprintln!("PopLast");
+                assert_eq!(std_btree.pop_last(), Some(entry.clone()));
+                assert_eq!(btree.pop_last(), Some(entry));
+            }
```
Comment (ielashi) [inline src/btreemap/proptests.rs:247]: I still don't understand why this isn't just:

```suggestion
            eprintln!("PopLast");
            assert_eq!(std_btree.pop_last(), btree.pop_last());
```

File: src/btreemap/proptests.rs [hunk 18]
```diff
@@ -231,5 +235,27 @@ fn execute_operation<M: Memory>(
...
+        }
+        Operation::PopFirst => {
+            assert_eq!(std_btree.len(), btree.len() as usize);
+
+            let idx = std_btree.len();
+
+            if let Some(entry) = btree.iter().skip(idx).take(1).next() {
```
Comment (ielashi) [inline src/btreemap/proptests.rs:254]: ditto

File: src/btreemap/proptests.rs [hunk 19]
```diff
@@ -231,5 +235,17 @@ fn execute_operation<M: Memory>(
                 assert_eq!(v1, &v2);
             }
         }
+        Operation::PopLast => {
+            assert_eq!(std_btree.len(), btree.len() as usize);
+
```
Comment (ielashi) [inline src/btreemap/proptests.rs:240]: nit: I guess this assertion is no longer relevant for this operation?

File: src/btreemap/proptests.rs [hunk 20]
```diff
@@ -231,5 +235,17 @@ fn execute_operation<M: Memory>(
...
+            eprintln!("PopLast");
+            assert_eq!(std_btree.pop_last(), btree.pop_last());
+        }
+        Operation::PopFirst => {
+            assert_eq!(std_btree.len(), btree.len() as usize);
```
Comment (ielashi) [inline src/btreemap/proptests.rs:245]: ditto

**Review by ielashi — APPROVED** [review]
Thanks for your contribution, Witter!
